server/src/session.rs

```diff
@@ -0,0 +1,120 @@
+        }
+    }
+
+    // Update the properties of the session from `InitializeParams`
+    pub async fn update_from_initialize_params(
+        &self,
+        params: tower_lsp::lsp_types::InitializeParams,
+    ) {
+        // This is the path to the built-in Slice files that are included with the extension. It should always
+        // be present.
+        let built_in_slice_path = params
+            .initialization_options
+            .and_then(|opts| opts.get("builtInSlicePath").cloned())
+            .and_then(|v| v.as_str().map(str::to_owned))
+            .expect("builtInSlicePath not found in initialization options");
+        *self.built_in_slice_path.write().await = built_in_slice_path;
+
+        // Use the root_uri if it exists temporarily as we cannot access configuration until
+        // after initialization. Additionally, LSP may provide the windows path with escaping or a lowercase
+        // drive letter. To fix this, we convert the path to a URL and then back to a path.
+        let root_uri = params
+            .root_uri
+            .and_then(|uri| uri.to_file_path().ok())
+            .and_then(|path| Url::from_file_path(path).ok())
+            .expect("root_uri not found in initialization parameters");
+        *self.root_uri.write().await = Some(root_uri);
+    }
+
+    // Update the stored configuration sets by fetching them from the client.
+    pub async fn fetch_configurations(&self) {
+        let built_in_path = &self.built_in_slice_path.read().await;
+        let root_uri_guard = self.root_uri.read().await;
+        let root_uri = (*root_uri_guard).clone().expect("root_uri not set");
+
+        let params = vec![ConfigurationItem {
+            scope_uri: None,
+            section: Some("slice.configurations".to_string()),
+        }];
+
+        // Fetch the configurations from the client and parse them.
+        let configurations = self
+            .client
+            .configuration(params)
+            .await
+            .ok()
+            .map(|response| {
+                let config_array = &response
+                    .iter()
+                    .filter_map(|config| config.as_array())
+                    .flatten()
+                    .cloned()
+                    .collect::<Vec<_>>();
+                ConfigurationSet::parse_configuration_sets(config_array, &root_uri, built_in_path)
+            })
+            .unwrap_or_default();
+
+        // Update the configuration sets
+        self.update_configurations(configurations).await;
+    }
+
+    // Update the configuration sets from the `DidChangeConfigurationParams` notification.
+    pub async fn update_configurations_from_params(&self, params: DidChangeConfigurationParams) {
+        let root_uri_guard = self.root_uri.read().await;
+        let root_uri = (*root_uri_guard).clone().expect("root_uri not set");
+        let built_in_path = &self.built_in_slice_path.read().await;
```

Comment:
... But in this function we lock in the opposite order. First grabbing `root_uri` then `built_in_slice_path`.

If these two functions run around the same time, the first function will grab `built_in_slice_path`, the 2nd function will grab `root_uri`, and then everything dies.

We must make sure to always lock these things in the same order.
That way function1 grabs `root_uri`, then (now that they're in the same order) function2 also tries to grab `root_uri`, and has to wait. Instead of being able to grab `built_in_slice_path`.